### ISSUE TITLE
Implement Automated Semantic Type Hierarchy from SVO Patterns

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#699]
## Overview

Implement fully automated semantic type discovery from corpus SVO patterns. **Zero human annotation required.**

This replaces the old M1 selectional preference model with a more efficient two-layer hybrid approach:
- **Layer 1**: Deterministic (0 params) - AST-based role extraction
- **Layer 2**: Statistical (500K params) - Semantic Fact Validator

**Expected**: 20x smaller than old M1 (500K vs 10M params), better interpretability

## Architecture

```
5.4M Sentences
    ↓
[Parser Enhancement #691] - Subordinate clauses → nested frazo nodes
    ↓
4.5-5.5M SVO Triples (+10-15% from subordinates)
    ↓
[Verb Co-occurrence Clustering] - Distributional hypothesis
    ↓
18 Semantic Type Clusters (PERSONO, ANIMALO, OBJEKTO, LOKO, etc.)
    ↓
[Selectional Preference Extraction]
    ↓
VERB_CONSTRAINTS.json - P(subject_type|verb), P(object_type|verb)
    ↓
[Semantic Fact Validator] - 500K param validation model
```

### ✅ Completed

1. **Parser Enhancement** (#691) - Closed
   - Nested frazo nodes for subordinate clauses
   - 12 subordinating conjunction types (ke, kiu, kiam, se, ĉar, etc.)
   - +10-15% SVO triple extraction improvement
   - Files: `klareco/parser.py`, `scripts/extract_svo_triples.py`

2. **Pipeline Scripts Created** (2026-03-16)
   - ✅ `scripts/cluster_semantic_types.py` - Verb co-occurrence clustering
   - ✅ `scripts/generate_verb_constraints.py` - Selectional preference extraction
   - ✅ `scripts/build_semantic_type_hierarchy.sh` - One-command pipeline
   - ✅ Documentation: `/tmp/semantic_type_hierarchy_guide.md`

### 🚧 In Progress

**Ready to execute** - All scripts implemented, need to run on corpus:

```bash
# Run full pipeline (5-9 hours)
./scripts/build_semantic_type_hierarchy.sh
```

### ⏳ Next Steps

1. **Run SVO Extraction** (4-8 hours)
   - Extract from 5.4M sentence corpus
   - Expected: 4.5-5.5M SVO triples
   - Output: `data/semantic_types/svo_triples_full.jsonl`

2. **Cluster Semantic Types** (10-30 minutes)
   - Hierarchical clustering on verb co-occurrence matrix
   - Expected: 18 semantic type clusters
   - Output: `data/semantic_types/semantic_types.json`

3. **Generate Verb Constraints** (5-10 minutes)
   - Compute P(subject_type|verb) and P(object_type|verb)
   - Apply Laplace smoothing
   - Output: `data/semantic_types/verb_constraints.json`

4. **Train Semantic Fact Validator** (TODO - #687 replacement)
   - Two-layer model: deterministic + 500K param statistical
   - Validates (subject_type, verb, object_type) plausibility
   - Replaces old M1 selectional preference model

5. **Integrate into RAG Pipeline** (TODO)
   - Use SFV to filter implausible facts during retrieval
   - Add to `klareco/rag/` pipeline

### Why This Approach?

**Old M1 Model** (#687):
- 10M parameters
- Learned subject-verb-object compatibility
- 80% accuracy but slow and opaque

**New Semantic Fact Validator**:
- 500K parameters (20x smaller)
- Two-layer hybrid (deterministic + statistical)
- Interpretable via semantic type labels
- Fully automated from corpus (no manual annotation)

### Distributional Hypothesis

> "Words appearing with similar verbs have similar semantic types"

Example:
- "zamenhof kreis", "napoleon kondukis", "einstein pensis" → PERSONO
- "hundo kurtas", "kato miautas", "birdo flugas" → ANIMALO
- "tablo staras", "seĝo estas", "libro kuŝas" → OBJEKTO

### Expected Semantic Types (18 clusters)

1. PERSONO - People
2. ANIMALO - Animals
3. OBJEKTO - Physical objects
4. LOKO - Places
5. TEMPO - Time
6. AGO - Actions/Events
7. ABSTRAKTO - Abstract concepts
8. KVALITO - Qualities
9. KVANTO - Quantities
10. TYPE_10 - TYPE_18: Auto-labeled clusters

## Outputs

```
data/semantic_types/
├── svo_triples_full.jsonl              # 4.5-5.5M SVO triples
├── semantic_types.json                 # Root → type mappings (~6K roots)
├── verb_constraints.json               # Verb selectional preferences
├── cluster_stats.json                  # Clustering quality metrics
├── constraint_stats.json               # Validation metrics
└── verb_cooccurrence_matrix.npz        # Sparse matrix (optional)
```

### Clustering Quality
- **Silhouette score**: >0.3 (good separation)
- **Within-cluster similarity**: >0.3 (coherent)
- **Coverage**: >95% of frequent roots

### Constraint Quality
- **Mean entropy**: 2.0-3.5 (balanced specificity)
- **Overfitting**: <5% of verbs
- **Fundamento coverage**: >90%

## Related Issues

- #691 - Parser Enhancement (subordinate clauses) - **CLOSED**
- #687 - Old M1 Selectional Preference (to be replaced)
- #625, #626, #627 - Entity type classifier (different approach)
- #696 - LLM-Style Capabilities Epic

## Documentation

- **Design**: `/tmp/automated_type_hierarchy_design.md`
- **Usage Guide**: `/tmp/semantic_type_hierarchy_guide.md`
- **Wiki**: [Parser-Subordinate-Clauses](https://github.com/marctjones/klareco/wiki/Parser-Subordinate-Clauses)
- **Release**: [v2.1-parser-enhancement](https://github.com/marctjones/klareco/releases/tag/v2.1-parser-enhancement)

### Scripts

All scripts include comprehensive docstrings with VERSION, COMPATIBLE WITH, DEPENDENCIES, STAGE.

**cluster_semantic_types.py**:
- Builds sparse verb co-occurrence matrix (roots × verbs)
- Hierarchical clustering with Ward linkage
- Auto-assigns semantic labels using heuristics
- Computes cluster quality metrics

**generate_verb_constraints.py**:
- Computes probability distributions P(type|verb)
- Applies Laplace smoothing (α=0.1)
- Calculates entropy (selectional specificity)
- Validates against overfitting

**build_semantic_type_hierarchy.sh**:
- One-command pipeline automation
- Handles checkpointing and resume
- Logging to `logs/semantic_type_hierarchy_*.log`

### Parameters

**Clustering**:
- `--num-clusters`: 18 (default) - number of semantic types
- `--min-frequency`: 10 - minimum root/verb frequency
- `--method`: ward/average/complete - linkage method

**Constraints**:
- `--min-frequency`: 5 - minimum verb frequency
- `--smoothing-alpha`: 0.1 - Laplace smoothing parameter

## Timeline

- **2026-03-16**: Scripts created, ready to run
- **Next**: Run pipeline on full corpus (~5-9 hours)
- **After**: Train Semantic Fact Validator model
- **Integration**: Add to RAG pipeline for fact filtering

## Acceptance Criteria

- [ ] SVO extraction completes successfully (4.5-5.5M triples)
- [ ] Clustering produces coherent semantic types (silhouette >0.3)
- [ ] Verb constraints cover >90% of Fundamento verbs
- [ ] Semantic Fact Validator trained (500K params)
- [ ] Integration into RAG pipeline working
- [ ] Performance evaluation vs old M1 model

## Labels

- `enhancement`
- `priority:P1`
- `type:learned`
- `training`
- `epic`

**Comment by marctjones** [comment]
## Architecture Update (2026-03-22)

After deep analysis, we're **simplifying the approach** based on AST-first principles:

### Original Plan (Complex)
- Extract semantic type hierarchy from corpus clustering
- Build VERB_CONSTRAINTS dictionary
- Train 500K param Semantic Fact Validator

### NEW Simplified Plan (RECOMMENDED)

**Just build a learned plausibility scorer - skip the complexity!**

**Architecture**:
- Frozen hybrid root embeddings (0 new params)
- Simple concatenation MLP scorer (**98K params** - 5x smaller!)
- Train on corpus positives + synthetic negatives
- No manual constraint extraction needed

**Why simpler is better**:
✅ **98K params** vs 500K (5x smaller!)
✅ Model learns optimal boundaries (no threshold tuning)
✅ Faster to implement (3-4 days vs weeks)
✅ Less brittle (no arbitrary thresholds)
✅ Still explainable via weight/embedding analysis

**Training Data**:
- Positive: 4.5M SVO triples from corpus (appeared in real text → plausible)
- Negative: 4.5M synthetic (role swaps, random combinations)

**Success Criteria**:
- Accuracy >85% on test set
- Precision/Recall >80% on implausible detection
- <100K parameters
- Trains in <3 hours on CPU

**Tasks Created**:
- Task #3: Generate training dataset (9M examples)
- Task #1: Build & train plausibility scorer (98K params)
- Task #4: Evaluate & integrate into AST pipeline
- Task #2: Analysis tool for explainability

**Related Design Docs**:
- `/tmp/semantic_fact_validator_design.md` - Full architectural analysis
- `/tmp/plausibility_scorer_design.md` - Detailed scorer design

This aligns better with VISION.md principles: minimize learned parameters, maximize simplicity, maintain explainability through analysis tools rather than complex deterministic layers.

**Comment by marctjones** [comment]
## Decision: Quality Over Quantity for Plausibility Training Dataset

**Changed approach from 9M to 200K quality-focused examples.**

### Rationale

1. **Parameter efficiency**: 98K param model only needs ~200K examples (2:1 ratio)
2. **Training speed**: 3-5 hours vs 6-9 days
3. **Quality focus**: Challenging negatives > simple random swaps
4. **Faster iteration**: Enable experimentation and error analysis

### Quality-Focused Approach

**Positive Examples (100K)**:
- Confidence threshold ≥ 0.9
- Balanced semantic type coverage
- Verb diversity (skip overly common verbs)

**Negative Examples (100K)**:
1. **Type-compatible swaps (40K)**: Same type, different word - semantically challenging
2. **Role confusion (30K)**: Real verb, wrong argument types - tests selectional preferences  
3. **Filtered random (30K)**: Edge cases, rare combinations

### Files Created

- `scripts/generate_plausibility_training_data_quality.py` - Quality-focused generation ✅
- `/tmp/plausibility_dataset_quality_decision.md` - Full decision rationale

### Next Steps

1. Extract SVO triples from ~500K sentences
2. Generate 200K quality training examples
3. Train plausibility scorer (Task #1)
4. Evaluate and iterate based on results

**Status**: Ready to proceed with quality-focused dataset generation.

**Comment by marctjones** [comment]
Closing as part of a 2026-09-08 backlog cleanup (see PLAN.md, DESIGN.md, CLAUDE.md's merge gate).

This issue predates the merge gate (adopted 2026-07-13) and the AST-native orchestrator rewrite. It describes work against the pre-orchestrator RAG pipeline or the learned model stack (M1/M2 embeddings, compositional embeddings, entity classifiers, neural rerankers) that was pruned from the repo -- klareco/embeddings/, klareco/models/, and klareco/summarization/ no longer exist. It has no milestone and was never benchmarked, matching exactly the failure mode CLAUDE.md documents (25 capability issues filed 2026-05-26, 8 implemented, none benchmarked).

Labeled archived-2026-09 so this batch is recoverable: gh issue list --state closed --label archived-2026-09. If this capability is still wanted, re-file it against the current architecture naming the benchmark metric and test the merge gate requires.
